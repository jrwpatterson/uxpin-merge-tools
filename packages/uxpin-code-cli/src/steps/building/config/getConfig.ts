// tslint:disable-next-line:import-name
import HappyPack = require('happypack');
import { join } from 'path';
import { Configuration } from 'webpack';
import { smart } from 'webpack-merge';

export const TEMP_DIR_PATH:string = './.uxpin-temp';
export const LIBRARY_INPUT_PATH:string = `${TEMP_DIR_PATH}/components.js`;
export const LIBRARY_OUTPUT_PATH:string = `${TEMP_DIR_PATH}/designsystemlibrary.js`;

export function getConfig(projectRoot:string, webpackConfigPath?:string):Configuration {
  const config:Configuration = {
    entry: LIBRARY_INPUT_PATH,
    module: {
      rules: [
        {
          exclude: /node_modules/,
          test: /.jsx?$/,
          use: 'happypack/loader',
        },
      ],
    },
    output: {
      filename: LIBRARY_OUTPUT_PATH,
      libraryTarget: 'commonjs',
    },
    plugins: [
      new HappyPack({
        loaders: [
          {
            loader: 'babel-loader',
            options: {
              plugins: [
                'transform-class-properties',
                'transform-object-rest-spread',
              ],
              presets: [
                ['env', {
                  targets: {
                    browsers: ['last 2 versions'],
                    node: 'current',
                  },
                }],
                'react',
              ],
            },
          },
        ],
      }),
    ],
    resolve: {
      extensions: ['.js', '.jsx'],
    },
  };

  if (webpackConfigPath) {
    const userWebpackConfig:Configuration = require(join(projectRoot, webpackConfigPath));
    return smart(userWebpackConfig, config);
  }
  return config;
}
