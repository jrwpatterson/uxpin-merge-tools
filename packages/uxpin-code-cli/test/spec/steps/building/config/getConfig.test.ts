import { Configuration } from 'webpack';
import { getConfig } from '../../../../../src/steps/building/config/getConfig';

describe('getConfig', () => {
  describe('returns webpack config merged with user config and enriched with happy pack config', () => {
    describe('when there is user wepback config and project uses babel with plugins', () => {
      it('returns correctly merged config', () => {
        // when
        const result:Configuration = getConfig();

        // then
        expect(result).toMatchSnapshot();
      });
    });
  });
});
