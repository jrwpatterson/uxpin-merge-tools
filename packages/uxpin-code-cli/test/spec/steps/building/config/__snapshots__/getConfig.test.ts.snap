// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig returns webpack config merged with user config and enriched with happy pack config when there is user wepback config and project uses babel with plugins returns correctly merged config 1`] = `
Object {
  "entry": "./.uxpin-temp/components.js",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\.jsx\\?\\$/,
        "use": "happypack/loader",
      },
    ],
  },
  "output": Object {
    "filename": "./.uxpin-temp/designsystemlibrary.js",
    "libraryTarget": "commonjs",
  },
  "plugins": Array [
    HappyPlugin {
      "config": Object {
        "compilerId": "default",
        "id": "1",
        "loaders": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "transform-class-properties",
                "transform-object-rest-spread",
              ],
              "presets": Array [
                Array [
                  "env",
                  Object {
                    "targets": Object {
                      "browsers": Array [
                        "last 2 versions",
                      ],
                      "node": "current",
                    },
                  },
                ],
                "react",
              ],
            },
          },
        ],
        "threads": 3,
        "verbose": true,
      },
      "id": "1",
      "name": "HappyPack",
      "state": Object {
        "baseLoaderRequest": "",
        "debug": false,
        "foregroundThreadPool": null,
        "loaders": Array [],
        "verbose": false,
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
  },
}
`;
